drools-wb-screens/drools-wb-dtable-xls-editor/drools-wb-dtable-xls-editor-backend/src/main/java/org/drools/workbench/screens/dtablexls/backend/server/conversion/GuidedDecisionTablePopulater.java

```diff
@@ -116,70 +118,131 @@ private void processIndirectSourceBuilders(final int maxRowCount) {
-                            column.getChildColumns().add(target);
-                            ivsIts.remove();
+                //Extract Template Keys for Patterns in iteration
+                boolean allPatternsHaveTemplateKeys = true;
+                for (int iPattern = previousPatternCount; iPattern < rm.lhs.length; iPattern++) {
+                    final IPattern pattern = rm.lhs[iPattern];
+                    final Map<InterpolationVariable, Integer> templateKeys = new HashMap<>();
+                    final RuleModelVisitor rmv = new RuleModelVisitor(templateKeys);
+                    rmv.visit(pattern);
+
+                    patternTemplateKeys.put(pattern, templateKeys);
+
+                    if (templateKeys.isEmpty()) {
+                        allPatternsHaveTemplateKeys = false;
+                    }
+                }
+
+                // Copy definition to column.
+                // If ALL patterns in the iteration have Template keys we can use IPattern otherwise fallback to FreeFormLine
+                if (allPatternsHaveTemplateKeys) {
+                    for (int iPattern = previousPatternCount; iPattern < rm.lhs.length; iPattern++) {
+                        final IPattern pattern = rm.lhs[iPattern];
+                        column.getDefinition().add(pattern);
+                    }
+                } else {
+                    final FreeFormLine ffl = new FreeFormLine();
+                    ffl.setText(lhsBuilders.get(i).getResult());
+                    column.setDefinition(Collections.singletonList(ffl));
+                }
+
+                //Copy variables to column
+                for (int iPattern = previousPatternCount; iPattern < rm.lhs.length; iPattern++) {
+                    final IPattern pattern = rm.lhs[iPattern];
+                    final Map<InterpolationVariable, Integer> templateKeys = patternTemplateKeys.get(pattern);
+                    if (Objects.nonNull(templateKeys)) {
+                        final List<InterpolationVariable> ivs = new ArrayList<>(templateKeys.keySet());
+                        for (BRLVariableColumn variableColumn : variableColumns) {
+                            final Iterator<InterpolationVariable> ivsIts = ivs.iterator();
+                            while (ivsIts.hasNext()) {
+                                final InterpolationVariable iv = ivsIts.next();
+                                if (iv.getVarName().equals(variableColumn.getVarName())) {
+                                    final BRLConditionVariableColumn source = (BRLConditionVariableColumn) variableColumn;
+                                    final BRLConditionVariableColumn target = makeBRLConditionVariableColumn(source, iv);
+                                    column.getChildColumns().add(target);
+                                    ivsIts.remove();
+                                }
+                            }
                         }
                     }
                 }
```

Comment:
Is that `final List<InterpolationVariable> ivs = new ArrayList<>(templateKeys.keySet());` really needed? I tried get rid of that line and tests passed. I also used `Objects.equals`

```java
                for (int iPattern = previousPatternCount; iPattern < rm.lhs.length; iPattern++) {
                    final IPattern pattern = rm.lhs[iPattern];
                    final Map<InterpolationVariable, Integer> templateKeys = patternTemplateKeys.get(pattern);
                    if (Objects.nonNull(templateKeys)) {
                        for (BRLVariableColumn variableColumn : variableColumns) {
                            final Iterator<InterpolationVariable> ivsIts = templateKeys.keySet().iterator();
                            while (ivsIts.hasNext()) {
                                final InterpolationVariable iv = ivsIts.next();
                                if (Objects.equals(iv.getVarName(), variableColumn.getVarName())) {
                                    final BRLConditionVariableColumn source = (BRLConditionVariableColumn) variableColumn;
                                    final BRLConditionVariableColumn target = makeBRLConditionVariableColumn(source, iv);
                                    column.getChildColumns().add(target);
                                    ivsIts.remove();
                                }
                            }
                        }
                    }
                }
```

Reply:
`final List<InterpolationVariable> ivs = new ArrayList<>(templateKeys.keySet())` is not as such needed (and your change works equally well) however I only need to call `templateKeys.keySet()` once (outside the `for (BRLVariableColumn variableColumn...` loop) whereas you're calling it inside it.. so consider it an optimisation!

I've changed to use `Objects.equals(...)` though (I know how you prefer it)